cpp/src/arrow/flight/sql/server_session_middleware.h

```diff
@@ -0,0 +1,88 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Middleware for handling Flight SQL Sessions including session cookie handling.
+// Currently experimental.
+
+#pragma once
+
+#include "arrow/flight/server_middleware.h"
+#include "arrow/flight/sql/types.h"
+
+namespace arrow {
+namespace flight {
+namespace sql {
+
+class ServerSessionMiddlewareFactory;
+
+static constexpr char const kSessionCookieName[] =
+    "flight_sql_session_id";
+
+class FlightSqlSession {
+ protected:
+  std::map<std::string, SessionOptionValue> map_;
```

Comment:
Does order matter? Can we use unordered_map?

Reply:
It doesn't but the decision was (consciously) made here (and I believe made by the STL designers when choosing what the more obvious 'map' type would be) on the basis that most of the tree-based std::map impls are faster than unordered_map (hash-based) for "up to around 10,000 entries".  (Performance issues with dense keys are also avoided by the former making the thing without unexpected degenerate performance cases the 'sane' default.)